### PR TITLE
intel_adsp_cavs15/ logtool.py: fix support for multiple NULLs

### LINKED ISSUE #34492: Logging still broken with SOF
**Description**
When built with Zephyr, SOF logs, as generated by the logtool, appear badly corrupted.

**To Reproduce**
Build and run SOF with Zephyr, print logs per
```
sudo tools/logtool.py -e sof
```

**Expected behavior**
Properly formatted consistent logs should be printed.

**Impact**
Debugging becomes impossible.

**Logs and console output**
Example output:
> 23086196: pipe reset
> 23092510: dai_get type 3 index 1 new sref 2
> 23093828: dai_put type 3 index 1 new sref 1
> d080002
> 23096241: dai_config() dai type = 3 index = 1
> 23097545: dai_config(), channel = -1
> ef 3
> 23111526: dai_get type 3 index 1 new sref 2
> x4001
> 23112817: dai_put type 3 index 1 new sref 1
> , sscr1 = 0xd0723115214: dai_config() dai type = 3 index = 1
> 00000, sscr3 = 0x07070000, ssioc = 0x00000020
> 23116497: dai_config(), channel = -1
> 0004002, sspsp2 = 0x00

**Context**

Logging configuration:

> CONFIG_LOG_BACKEND_ADSP=y
> CONFIG_LOG=y
> CONFIG_SOC_LOG_LEVEL_INF=y
> CONFIG_SOC_LOG_LEVEL=3
> CONFIG_ARCH_LOG_LEVEL_INF=y
> CONFIG_ARCH_LOG_LEVEL=3
> CONFIG_MPU_LOG_LEVEL_INF=y
> CONFIG_MPU_LOG_LEVEL=3
> CONFIG_KERNEL_LOG_LEVEL_INF=y
> CONFIG_KERNEL_LOG_LEVEL=3
> CONFIG_IPM_LOG_LEVEL_INF=y
> CONFIG_IPM_LOG_LEVEL=3
> CONFIG_LOG_MODE_DEFERRED=y
> CONFIG_LOG_DEFAULT_LEVEL=3
> CONFIG_LOG_OVERRIDE_LEVEL=0
> CONFIG_LOG_MAX_LEVEL=4
> CONFIG_LOG_FUNC_NAME_PREFIX_DBG=y
> CONFIG_LOG_MODE_OVERFLOW=y
> CONFIG_LOG_PROCESS_TRIGGER_THRESHOLD=10
> CONFIG_LOG_PROCESS_THREAD=y
> CONFIG_LOG_PROCESS_THREAD_SLEEP_MS=1000
> CONFIG_LOG_PROCESS_THREAD_STACK_SIZE=1024
> CONFIG_LOG_BUFFER_SIZE=1024
> CONFIG_LOG_DETECT_MISSED_STRDUP=y
> CONFIG_LOG_STRDUP_MAX_STRING=32
> CONFIG_LOG_STRDUP_BUF_COUNT=4
> CONFIG_LOG_DOMAIN_ID=0

### PR DESCRIPTION
**Pull request by marc-hb**
Stop printing garbage when strings are terminated with more than one
zero byte.

Thanks Andy for the diagnosis (and the better adsplog.py)

Fixes #34492

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

### CODE REVIEW

**Review by finikorg — APPROVED** [review]

**Review by lyakh — APPROVED** [review]

**Comment by marc-hb** [comment]
Note #35186 wants to deletes and replace this script entirely for 2.7. I still think this PR here is a valuable one-line fix in the shorter 2.6 term. I heard this older script is used in some automation somewhere.

**Comment by andyross** [comment]
Some background: The current Zephyr trace backend will terminate (i.e. fill in the magic number and a correct sequential ID) and push out a null-terminated slot containing less than 60 printable characters at the end of a single write (generally a single LOG_*() or printk() call) so that it can be seen synchronously by the client.  The older code would do this too, I think, but it would IIRC leave that last slot in place and keep adding to it, which is racy if you're trying to read the output in realtime and not in a single pass.  (There is also some complexity with ordering the writes to the magic number and ID such that the kernel can't get confused about the state of slot zero, but I don't think that's relevant to this particular situation.)

Also, it seems like older code might have kept the slot zero-filled (which is what this older script seems to be expecting), where the current backend will only write bytes up to the null and leave the rest with whatever content they had from the last write to the slot.

Basically: this is what happens when you have a protocol that isn't expressly specified.  It's not unlikely that the trace code in the SOF kernel driver has different edge cases too. :)

FWIW: as mentioned, another way to address this same issue would be to zero-out the whole slot in the firmware.  Note here where all it does is write an initial zero to the new slot: https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/xtensa/intel_adsp/common/trace_out.c#L117

**Review by andyross — APPROVED** [review]
No reason not to merge this though.  It fixes a bug and the removal PR can't possibly merge before 2.6 ships.

**Comment by andyross** [comment]
Also, in the longer term (especially since dictionary logging is arriving which will change the intermediate protocol and require kernel changes anyway) we might consider stepping back and revisiting this protocol.   A simpler ring buffer would definitely be better, as would DMA transport for the data.  I know @marc-hb has been looking at the latter, but we could actually do both at the same time (and even in the same buffer -- read it simply from a userspace python script for debugging but DMA simultaneously into the kernel for production use).

**Review by dcpleung — APPROVED** [review]
